File: AtomykPlay/app/src/main/java/com/atomykcoder/atomykplay/fragments/FavoritesFragment.java
package com.atomykcoder.atomykplay.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.atomykcoder.atomykplay.R;
import com.atomykcoder.atomykplay.adapters.FavoriteListAdapter;
import com.atomykcoder.atomykplay.adapters.SimpleTouchCallback;
import com.atomykcoder.atomykplay.customScripts.LinearLayoutManagerWrapper;
import com.atomykcoder.atomykplay.data.Music;
import com.atomykcoder.atomykplay.events.RemoveFromFavoriteEvent;
import com.atomykcoder.atomykplay.utils.StorageUtil;
import com.atomykcoder.atomykplay.interfaces.OnDragStartListener;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.util.ArrayList;

public class FavoritesFragment extends Fragment implements OnDragStartListener {

    private ItemTouchHelper itemTouchHelper;
    private FavoriteListAdapter playListAdapter;
    private StorageUtil storageUtil;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_favorites, container, false);

        if (!EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().register(this);
        }

        storageUtil = new StorageUtil(requireContext());

        RecyclerView recyclerView = view.findViewById(R.id.favorite_music_recycler);
        View noPlLayout = view.findViewById(R.id.song_not_found_layout_favorite);
        Toolbar toolbar = view.findViewById(R.id.toolbar_favorites);
        toolbar.setNavigationIcon(R.drawable.ic_back);
        toolbar.setNavigationOnClickListener(v -> requireActivity().onBackPressed());

        recyclerView.setHasFixedSize(true);

        LinearLayoutManager manager = new LinearLayoutManagerWrapper(getContext());

        ArrayList<Music> favList = storageUtil.getFavouriteList();


        if (favList != null) {
            playListAdapter = new FavoriteListAdapter(requireContext(), favList, this);
            recyclerView.setLayoutManager(manager);
            recyclerView.setAdapter(playListAdapter);
            noPlLayout.setVisibility(View.GONE);

            if (favList.isEmpty()) {
                noPlLayout.setVisibility(View.VISIBLE);
            }

            ItemTouchHelper.Callback callback = new SimpleTouchCallback(playListAdapter);
            itemTouchHelper = new ItemTouchHelper(callback);
            itemTouchHelper.attachToRecyclerView(recyclerView);
        } else {
            noPlLayout.setVisibility(View.VISIBLE);

        }

        return view;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
    }

    @Override
    public void onDragStart(RecyclerView.ViewHolder viewHolder) {
        itemTouchHelper.startDrag(viewHolder);
    }

    @Subscribe
    public void removeFromPlaylist(RemoveFromFavoriteEvent event) {
        playListAdapter.removeItem(event.music);
    }
}

File: AtomykPlay/app/src/main/java/com/atomykcoder/atomykplay/fragments/FavoritesFragment.kt
package com.atomykcoder.atomykplay.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.atomykcoder.atomykplay.R
import com.atomykcoder.atomykplay.adapters.FavoriteListAdapter
import com.atomykcoder.atomykplay.adapters.SimpleTouchCallback
import com.atomykcoder.atomykplay.data.BaseFragment
import com.atomykcoder.atomykplay.interfaces.OnDragStartListener
import com.atomykcoder.atomykplay.models.Music
import com.atomykcoder.atomykplay.scripts.LinearLayoutManagerWrapper
import com.atomykcoder.atomykplay.state.StateHolder
import com.atomykcoder.atomykplay.utils.StorageUtil
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.flow.collectLatest

class FavoritesFragment : BaseFragment(), OnDragStartListener {
    companion object {
        @JvmStatic
        fun newInstance() = FavoritesFragment()
    }

    private var itemTouchHelper: ItemTouchHelper? = null
    private var playListAdapter: FavoriteListAdapter? = null
    private var storageUtil: StorageUtil? = null
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_favorites, container, false)
        storageUtil = StorageUtil(requireContext())
        val recyclerView = view.findViewById<RecyclerView>(R.id.favorite_music_recycler)
        val noPlLayout = view.findViewById<View>(R.id.song_not_found_layout_favorite)
        val toolbar = view.findViewById<Toolbar>(R.id.toolbar_favorites)
        toolbar.setNavigationIcon(R.drawable.ic_back)
        toolbar.setNavigationOnClickListener { requireActivity().onBackPressed() }
        recyclerView.setHasFixedSize(true)
        val manager: LinearLayoutManager = LinearLayoutManagerWrapper(context)
        val favList: ArrayList<Music> = storageUtil!!.favouriteList

        playListAdapter = FavoriteListAdapter(requireContext(), favList, this)
        recyclerView.layoutManager = manager
        recyclerView.adapter = playListAdapter
        noPlLayout.visibility = View.GONE
        if (favList.isEmpty()) {
            noPlLayout.visibility = View.VISIBLE
        }
        val callback: ItemTouchHelper.Callback = SimpleTouchCallback(playListAdapter)
        itemTouchHelper = ItemTouchHelper(callback)
        itemTouchHelper!!.attachToRecyclerView(recyclerView)

        // Keep the list in sync with FavoriteStateManager - e.g. when a song is removed
        // from favorites via the option sheet elsewhere in the app while this list is open.
        viewLifecycleOwner.lifecycleScope.launchWhenStarted {
            StateHolder.favoriteStateManager.favoriteIds.collectLatest { ids ->
                playListAdapter?.removeItemsNotIn(ids)
            }
        }
        return view
    }

    override fun onDragStart(viewHolder: RecyclerView.ViewHolder) {
        itemTouchHelper!!.startDrag(viewHolder)
    }

}